sdk/cosmos/azure-cosmos-spark_3/src/main/scala/com/azure/cosmos/spark/TransactionalBulkWriter.scala

```diff
@@ -0,0 +1,1570 @@
+                s"for item ${itemOperation.getId}, PK: ${itemOperation.getPartitionKeyValue}")
+            }
+          }
+        }
+        finally {
+          if (!isGettingRetried.get) {
+            semaphore.release()
+          }
+        }
+
+        if (!shouldSkipTaskCompletion.get) {
+          markTaskCompletion()
+        }
+      },
+      errorConsumer = Option.apply(
+        ex => {
+          log.logError(s"Unexpected failure code path in Bulk ingestion, " +
+            s"Context: ${operationContext.toString} $getThreadInfo", ex)
+          // if there is any failure this closes the bulk.
+          // at this point bulk api doesn't allow any retrying
+          // we don't know the list of failed item-operations
+          // they only way to retry to keep a dictionary of pending operations outside
+          // so we know which operations failed and which ones can be retried.
+          // this is currently a kill scenario.
+          captureIfFirstFailure(ex)
+          cancelWork()
+          markTaskCompletion()
+        }
+      )
+    )
+  }
+
+  override def scheduleWrite(partitionKeyValue: PartitionKey, objectNode: ObjectNode): Unit = {
+    Preconditions.checkState(!closed.get())
+    throwIfCapturedExceptionExists()
+
+    val activeTasksSemaphoreTimeout = 10
+    val operationContext = new OperationContext(
+      getId(objectNode),
+      partitionKeyValue,
+      getETag(objectNode),
+      1,
+      monotonicOperationCounter.incrementAndGet(),
+      None)
+    val numberOfIntervalsWithIdenticalActiveOperationSnapshots = new AtomicLong(0)
+    // Don't clone the activeOperations for the first iteration
+    // to reduce perf impact before the Semaphore has been acquired
+    // this means if the semaphore can't be acquired within 10 minutes
+    // the first attempt will always assume it wasn't stale - so effectively we
+    // allow staleness for ten additional minutes - which is perfectly fine
+    var activeBulkWriteOperationsSnapshot = mutable.Set.empty[CosmosItemOperation]
+    var pendingBulkWriteRetriesSnapshot = mutable.Set.empty[CosmosItemOperation]
+    var activeReadManyOperationsSnapshot = mutable.Set.empty[ReadManyOperation]
+    var pendingReadManyRetriesSnapshot = mutable.Set.empty[ReadManyOperation]
+
+    log.logTrace(
+      s"Before TryAcquire ${totalScheduledMetrics.get}, Context: ${operationContext.toString} $getThreadInfo")
+    while (!semaphore.tryAcquire(activeTasksSemaphoreTimeout, TimeUnit.MINUTES)) {
+      log.logDebug(s"Not able to acquire semaphore, Context: ${operationContext.toString} $getThreadInfo")
+      if (subscriptionDisposable.isDisposed ||
+          (readManySubscriptionDisposableOpt.isDefined && readManySubscriptionDisposableOpt.get.isDisposed)) {
+        captureIfFirstFailure(
+          new IllegalStateException("Can't accept any new work - BulkWriter has been disposed already"))
+      }
+
+      throwIfProgressStaled(
```

Comment:
I think all the Semaphore dance should only start when you formed a batch - no need to do this in the middle of a batch. Semantically TransactionalBulkWriter should limit the number of outstanding batches - not operations.